nflow-engine/src/main/java/io/nflow/engine/internal/guice/EngineModule.java

```diff
@@ -0,0 +1,176 @@
+import org.springframework.transaction.support.TransactionTemplate;
+
+import javax.inject.Inject;
+import javax.sql.DataSource;
+import java.util.concurrent.ThreadFactory;
+
+public class EngineModule extends AbstractModule {
+
+  private final Object metricRegistry;
+  private final EngineConfiguration engineConfiguration;
+
+  public EngineModule(final Object metricRegistry, final EngineConfiguration engineConfiguration) {
+    this.metricRegistry = metricRegistry;
+    this.engineConfiguration = engineConfiguration;
+  }
+
+  @Override
+  protected void configure() {
+    TransactionSynchronizationManager.initSynchronization();
+    TransactionSynchronizationManager.setActualTransactionActive(true);
+    install(new EngineInitModule());
+  }
+
+  @Provides
+  @Inject
+  @NFlow
+  @Singleton
+  public AbstractResource nflowNonSpringWorkflowsListing(Environment env) {
+    return engineConfiguration.nflowNonSpringWorkflowsListing(env);
+  }
+
+  @Provides
+  @NFlow
+  @Singleton
+  public ThreadFactory nflowThreadFactory() {
+    return engineConfiguration.nflowThreadFactory();
+  }
+
+  @Provides
+  @Inject
+  @Singleton
+  public WorkflowInstanceExecutor nflowExecutor(@NFlow ThreadFactory factory, Environment env) {
+    return engineConfiguration.nflowExecutor(factory, env);
+  }
+
+  @Provides
+  @NFlow
+  @Singleton
+  public ObjectMapper nflowObjectMapper() {
+    return engineConfiguration.nflowObjectMapper();
+  }
+
+  @Provides
+  @NFlow
+  @Singleton
+  public DataSource nflowDataSource(Environment env) {
+    return getDatabaseConfiguration(env).nflowDatasource(env, metricRegistry);
+  }
+
+  @Provides
+  @NFlow
+  @Singleton
+  @Inject
+  public DatabaseInitializer nflowDatabaseInitializer(@NFlow DataSource dataSource, Environment env) {
+    return getDatabaseConfiguration(env).nflowDatabaseInitializer(dataSource, env);
+  }
+
+  @Provides
+  @NFlow
+  @Singleton
+  @Inject
+  public JdbcTemplate nflowJdbcTemplate(@NFlow DataSource dataSource, Environment env) {
+    return getDatabaseConfiguration(env).nflowJdbcTemplate(dataSource);
+  }
+
+  @Provides
+  @NFlow
+  @Singleton
+  @Inject
+  public NamedParameterJdbcTemplate nflowNamedParameterJdbcTemplate(@NFlow DataSource dataSource, Environment env) {
+    return getDatabaseConfiguration(env).nflowNamedParameterJdbcTemplate(dataSource);
+  }
+
+  @Provides
+  @NFlow
+  @Singleton
+  @Inject
+  public TransactionTemplate nflowTransactionTemplate(@NFlow DataSource dataSource, Environment env) {
+    return getDatabaseConfiguration(env).nflowTransactionTemplate(new DataSourceTransactionManager(dataSource));
+  }
+
+  @Provides
+  @Singleton
+  @Inject
+  public SQLVariants nflowSQLVariants(Environment env) {
+    return getDatabaseConfiguration(env).sqlVariants();
+  }
+
+  private DatabaseConfiguration getDatabaseConfiguration(Environment env) {
+    DatabaseConfiguration db;
+    String dbtype = env.getProperty("nflow.db.type", String.class);
+    switch (dbtype) {
```

Comment:
This throws nullptr if `nflow.db.type` is not defined -> should we default to `""` or `"h2"`?